db/db_iter.cc

```diff
@@ -456,7 +456,7 @@ bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check) {
       return false;
     }
 
-    if (iterate_upper_bound_ != nullptr &&
+    if (iterate_upper_bound_ != nullptr && !iter_->HintWithinUpperBound() &&
```

Comment:
We should think about improving the naming of `HintWithinUpperBound()`. Can it be called like `MaybeOutOfBound()` or something like that?

Reply:
My English is not good enough to get a good name. When the function returns true, `MaybeOutOfBound` suggests there's chance the key is not out of bound; When the function returns false, `IsOutOfBound` suggests the key is definitely not out of bound. Neither of them accurately describe the intent of the return value.

Reply:
MaybeOutOfBound() = true means it can be out-of-bound or in-bound, so it cannot skip upper bound check.  If it is = false, it means it cannot be out of bound, so we don't have to check.
I add the name `Maybe` because it is the one used in bloom filter MayContains() = true means it can contain or not contain. MayContains() = false means that it must not contain.
It's not a great one either, but HintWithinUpperBound() doesn't feel right either.

Reply:
ah, I misread it. Maybe `MayBe` is better than `Maybe` :) Will update after the other patch landed.